mcstatus/protocol/connection.py

```diff
@@ -231,15 +231,22 @@ def __del__(self):
             pass
 
 
+class NoConnectionsPossibleError(Exception):
+    pass
+
+
 class UDPSocketConnection(Connection):
     def __init__(self, addr: Address, timeout: float = 3):
         Connection.__init__(self)
         self.addr = addr
-        self.socket = socket.socket(
-            socket.AF_INET if ip_type(addr[0]) == 4 else socket.AF_INET6,
-            socket.SOCK_DGRAM,
-        )
-        self.socket.settimeout(timeout)
+        possible_connections = socket.getaddrinfo(addr.host, addr.port, type=socket.SOCK_DGRAM)
+        for conn in possible_connections:
+            self.socket = socket.socket(*conn[0:2])
+            self.socket.settimeout(timeout)
+            break
+        else:
+            raise NoConnectionsPossibleError
+        self.socket.connect(conn[-1])
```

Comment:
This logic is very odd, why not just access the first tuple in the list of `possible_connection` and except `IndexError`?

```suggestion
        possible_connections = socket.getaddrinfo(addr.host, addr.port, type=socket.SOCK_DGRAM)
        try:
            conn = possible_connections[0]
        except IndexError:
            raise NoConnectionPossibleError
        else:
            self.socket = socket.socket(*conn[0:2])
            self.socket.settimeout(timeout)
        self.socket.connect(conn[-1])
```

Reply:
Yeah, I was wondering if I should do a random index here instead. This function is retried multiple times, and I don't know if getaddrinfo is sorting the values that it returns.

Reply:
Well, the implementation you went with doesn't do that anyway, you could do `random,choice`, but I think it's fine to just go with the first result, if a random result can be expected to work, so should the first one and it's faster to pick nth element than to pick a random one.

Reply:
I'll write some tests for this bit and steer towards selecting a random choice. This way IPv6 and load balancers stand a chance.

Reply:
Output from `socket.getaddrinfo` isn't sorted in any way, that means that calling it twice likely won't give you the same order anyway, doing a random choice on randomly ordered set is pointless.

At least this is the case for IPv4 addresses, I haven't tested whether IPv6 ones would be preferred/unpreferred but I highly doubt they would.

Reply:
Oh, that's interesting, from my testing the returned order was always different, so I just assumed this was the case every time. I suppose pihole/unbound must be randomizing them.

> It would be more elegant to hoist the address resolution to be available before the decorated function is invoked. Then, it would be possible to pass a list of parameter tuples for each subsequent retry. This would allow trying each address at least once; rather than hoping that N random retries will eventually select a good address.

I think we should go with this route, even though it may be a bit more complex to implement. With just 3 retries and 2 IPs out of which 1 isn't working, there's still a 12.5% chance we pick the same one every time, which isn't ideal. Blindly retrying probably shouldn't be the way to go here.

We could probably make a generator function in `Address` class, or in the connection classes to give out these IPs and simply get the next one if the current one isn't working